Modernise lua-archive

hadess opened the pull request:
Lua 5.2 and 5.3 support, as well as porting to libarchive 3.0.

Untested.

hadess committed:
Add support for Lua 5.2

Use lua_getuservalue()/lua_setuservalue() instead of the removed
lua_getfenv()/lua_setfenv().

hadess committed:
Port to libarchive 3.0

Replace the deprecated functions with their libarchive 3.0 replacements
as listed at:
https://github.com/libarchive/libarchive/wiki/ManPageLibarchiveChanges3

hadess committed:
Fix use of unset variable

"wrote" was checked against -1 but was never actually set.

brimworks commented on ar.c:
```diff
@@ -38,8 +38,12 @@ static int ar_version(lua_State *L) {
 }
 
 //////////////////////////////////////////////////////////////////////
+#if !defined LUA_VERSION_NUM
```
Can this #if be refactored into a header (perhaps ar.h?) so we can avoid copying it into ar_entry.c?

hadess replied:
Sure.

brimworks commented on CMakeLists.txt:
```diff
@@ -9,6 +9,8 @@
 PROJECT(lua-archive C)
 CMAKE_MINIMUM_REQUIRED (VERSION 2.6)
 
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O0 -g3 -ggdb")
```
Perhaps this should be reverted since production installs will probably not want this setting? ...or at least comment it out?

hadess replied:
Sure, will see what the minimum needed is to have the CFLAGS actually applied.

brimworks commented on ar_read.c:
```diff
@@ -113,12 +113,16 @@ static int ar_read(lua_State *L) {
     if ( ! lua_isfunction(L, -1) ) err("MissingArgument: required parameter 'reader' must be a function");
     lua_setfield(L, -2, "reader"); // {ud}, {fenv}
-    lua_setfenv(L, -2); // {ud}
+#if LUA_VERSION_NUM > 501
```
Can this be refactored into ar.h (the new header I mentioned) so it looks something like this instead?

```
#if LUA_VERSION_NUM <= 501
#    define lua_setuservalue lua_setfenv
#endif
```

hadess replied:
Will do. Note that I'm still working on this as the test suite crashes. There were some other changes in the Lua sandboxing that might be causing this. Still investigating.

hadess committed:
Fix invalid FFlag in tests

When running the tests, libarchive says:
InvalidFFlag: 'archive' is not a known fflag

So remove that one flag from our tests.

hadess committed:
Add support for Lua 5.1

From http://lua-users.org/wiki/CompatibilityWithLuaFive:
luaL_reg (Lua 5.0) changed its name to luaL_Reg since Lua 5.1.

Apply the recommended fix to support lua 5.1 and prior versions.

hadess commented:
Note that I had to disable the "__gc" functions for the code to work and the tests to pass on Lua 5.3.

I don't think that the existing code should regress on 5.1 or older, but I'm still investigating the problems with newer Lua versions.

tomprince commented:
It at least compiles and loads for me.
